### PR TITLE
WIP: QuickBundlesX

### PR DESCRIPTION
**Pull request by Garyfallidis**
Hi all, 

This is the Hierarchical version of QuickBundles.

I still need to update the tests in this PR. But getting closer and closer.
This PR needs #1358 to be merged first. Please look into it.

More asap!

### CODE REVIEW

**Comment by arokem** [comment]
This PR adds more than 2,000 lines of code. That's a mammoth job to review. Any chance this could be broken down to a few (3?) smaller PRs? Thanks!

File: dipy/workflows/align.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
+from __future__ import division, print_function, absolute_import
+import logging
+from dipy.align.reslice import reslice
+from dipy.io.image import load_nifti, save_nifti
+from dipy.workflows.workflow import Workflow
+
+
+class ResliceFlow(Workflow):
```
Comment (arokem) [inline dipy/workflows/align.py]: For example, does this really belong in this PR? Seems unrelated to quickbundles... Maybe separate this off to another PR?

**Comment by Garyfallidis** [comment]
Hi @arokem this PR looks very big but it is not. We just need to merge #1358 first. And then will rebase.

**Comment by Garyfallidis** [comment]
The key files here are clustering_algorithms.pyx, clusteringspeed.pyx (and .pxd) and clustering.py

**Comment by Garyfallidis** [comment]
So basically only the last 8 commits are relevant.

**Comment by arokem** [comment]
Great!

File: dipy/segment/clustering.py [hunk 2]
```diff
@@ -508,3 +508,339 @@ def cluster(self, streamlines, ordering=None):
...
+    ----------
+    .. [Garyfallidis12] Garyfallidis E. et al., QuickBundles a method for
+                        tractography simplification, Frontiers in Neuroscience,
+                        vol 6, no 175, 2012.
+    """
+    def __init__(self, threshold, metric="MDF_12points",
+                 max_nb_clusters=np.iinfo('i4').max, bvh=False):
+        self.threshold = threshold
+        self.max_nb_clusters = max_nb_clusters
+        self.bvh = bvh
```
Comment (MarcCote) [inline dipy/segment/clustering.py]: I don't remember what `bvh`means?

File: dipy/segment/clustering.py [hunk 3]
```diff
@@ -508,3 +508,339 @@ def cluster(self, streamlines, ordering=None):
 
         cluster_map.refdata = streamlines
         return cluster_map
+
+
+class QuickBundlesOnline(Clustering):
+    r""" Clusters streamlines using QuickBundles [Garyfallidis12]_.
```
Comment (MarcCote) [inline dipy/segment/clustering.py]: This docstring needs to be updated to reflect the "online" aspect of that algorithm.

File: dipy/segment/clustering.py [hunk 4]
```diff
@@ -508,3 +508,339 @@ def cluster(self, streamlines, ordering=None):
...
+
+    @property
+    def clusters(self):
+        if self._dirty:
+            self._clusters = self._qb_state.get_cluster_map()
+
+        return self._clusters
+
+    def cluster(self, streamline, idx=None):
+        """ Clusters `streamlines` into bundles.
```
Comment (MarcCote) [inline dipy/segment/clustering.py]: This method clusters **one** streamline at the time because this is the online version of QB.

File: dipy/segment/clustering.py [hunk 5]
```diff
@@ -508,3 +508,339 @@ def cluster(self, streamlines, ordering=None):
...
+        def _traverse(node, level=0):
+            if level == wanted_level:
+                clusters.add_cluster(node)
+                return
+
+            for child in node.children:
+                _traverse(child, level+1)
+
+        _traverse(self.root)
+        return clusters
```
Comment (MarcCote) [inline dipy/segment/clustering.py]: Missing newline

File: dipy/segment/clustering_algorithms.pyx [hunk 6]
```diff
@@ -106,3 +114,129 @@ def quickbundles(streamlines, Metric metric, double threshold, long max_nb_clust
...
+        # The update step is performed right after the assignement step instead
+        # of after all streamlines have been assigned like k-means algorithm.
+        qb.update_step(cluster_id)
+        return qb, cluster_id
+
+    return _step
+
+
+def quickbundlesX(streamlines, Metric metric, thresholds, ordering=None):
+    """ Clusters streamlines using QuickBundles.
```
Comment (MarcCote) [inline dipy/segment/clustering_algorithms.pyx]: QB->QBx

File: dipy/segment/clusteringspeed.pxd [hunk 7]
```diff
@@ -40,8 +80,35 @@ cdef class QuickBundles(object):
...
+    cdef StreamlineInfos* current_streamline
+
+    cdef int _add_child(self, CentroidNode* node) nogil
+    cdef void _update_node(self, CentroidNode* node, StreamlineInfos* streamline_infos) nogil
+    cdef void _insert_in(self, CentroidNode* node, StreamlineInfos* streamline_infos, int[:] path) nogil
+    cpdef object insert(self, Data2D datum, int datum_idx)
+    cdef void traverse_postorder(self, CentroidNode* node, void (*visit)(QuickBundlesX, CentroidNode*))
+    cdef void _dealloc_node(self, CentroidNode* node)
+    cdef void _fetch_level(self, CentroidNode* node)
+    cdef object _build_tree_clustermap(self, CentroidNode* node)
```
Comment (MarcCote) [inline dipy/segment/clusteringspeed.pxd]: newline

File: dipy/segment/clusteringspeed.pxd [hunk 8]
```diff
@@ -2,14 +2,55 @@ from cythonutils cimport Data2D, Shape, shape2tuple, tuple2shape
...
+    float[6] aabb
 
 
 cdef struct NearestCluster:
     int id
     double dist
+    int flip
+
+
+cdef struct Test:
```
Comment (MarcCote) [inline dipy/segment/clusteringspeed.pxd]: Not sure that is still used.

File: dipy/segment/clusteringspeed.pyx [hunk 9]
```diff
@@ -1,18 +1,555 @@
...
+
+
+cdef CentroidNode* create_empty_node(Shape centroid_shape, float threshold) nogil:
+    # Important: because the CentroidNode structure contains an uninitialized memview,
+    # we need to zero-initialize the allocated memory (calloc or via memset),
+    # otherwise during assignment CPython will try to call _PYX_XDEC_MEMVIEW on it and segfault.
+    cdef CentroidNode* node = <CentroidNode*> calloc(1, sizeof(CentroidNode))
+    with gil:
+        node.centroid = <float[:centroid_shape.dims[0], :centroid_shape.dims[1]]> calloc(centroid_shape.size, sizeof(float))
+        #node.updated_centroid = <float[:centroid_shape.dims[0], :centroid_shape.dims[1]]> calloc(centroid_shape.size, sizeof(float))
```
Comment (MarcCote) [inline dipy/segment/clusteringspeed.pyx]: unused

File: dipy/segment/clusteringspeed.pyx [hunk 10]
```diff
@@ -1,18 +1,555 @@
...
+
+            # Build centroid
+            for n in range(self.root.centroid.shape[0]):
+                for d in range(self.root.centroid.shape[1]):
+                    self.root.centroid[n, d] = streamline.features[n, d]
+
+            # Build AABB
+            aabb_creation(self.root.centroid, self.root.aabb)
+
+            print "Done"
```
Comment (MarcCote) [inline dipy/segment/clusteringspeed.pyx]: Remove debug prints

File: dipy/segment/clusteringspeed.pyx [hunk 11]
```diff
@@ -1,18 +1,555 @@
...
+
+            # Insert the streamline in this new tree
+            #print "Try reinserting..."
+            self._insert(streamline)
+
+    cpdef void insert(self, Data2D datum, int datum_idx):
+        self.metric.feature.c_extract(datum, self.features)
+        self.metric.feature.c_extract(datum[::-1], self.features_flip)
+
+        # Important: because the CentroidNode structure contains an uninitialized memview,
```
Comment (MarcCote) [inline dipy/segment/clusteringspeed.pyx]: CentroidNode -> StreamlineInfos

File: dipy/segment/clusteringspeed.pyx [hunk 12]
```diff
@@ -314,3 +885,39 @@ cdef class QuickBundles(object):
...
+            cluster.indices = np.asarray(<int[:self.clusters.clusters_size[k]]> self.clusters.clusters_indices[k]).copy()
+            clusters.add_cluster(cluster)
+
+        return clusters
+
+    def get_cluster_map(self):
+        return self._build_clustermap()
+
+
+def evaluate_aabbb_checks():
```
Comment (MarcCote) [inline dipy/segment/clusteringspeed.pyx]: Is this used somewhere? If so, there's a typo in the name -> evaluate_aa**b**bb_checks

File: dipy/segment/tests/test_qbx.py [hunk 13]
```diff
@@ -0,0 +1,335 @@
...
+                #    hls[1] += 0.3
+
+                rbg = colorsys.hls_to_rgb(hls[0], luminosity, hls[2])
+                _color_subtree(child, np.asarray(rbg), level+1)
+
+    _color_subtree(tree.root)
+
+
+def test_show_qbx_tree():
+    filename = "/home/marc/research/dat/streamlines/MPI_Camille/myBrain.trk"
```
Comment (MarcCote) [inline dipy/segment/tests/test_qbx.py]: Need to fix the tests

File: dipy/segment/tests/test_qbx.py [hunk 14]
```diff
@@ -0,0 +1,335 @@
...
+        clusters = tree.get_clusters(level)
+        clusters.refdata = streamlines
+        show_clusters(clusters)
+
+
+if __name__ == '__main__':
+    #test_with_simulated_bundles2()
+    #test_show_qbx_tree()
+    #test_show_qbx()
+    test_3D_segments()
```
Comment (MarcCote) [inline dipy/segment/tests/test_qbx.py]: newline

File: dipy/segment/clustering.py (hunk 2 above)
Comment (Garyfallidis) [reply dipy/segment/clustering.py]: bounding volume hierarchy

**Comment by Garyfallidis** [comment]
Closed PR in favor of #1434
